File: Data Structures & Algorithms/count-number-of-islands/submission-2.cpp
class Solution {
public:
    int res = 0;
    int d[4][2] = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
    vector<vector<char>> mp;
    int m;
    int n;
    int numIslands(vector<vector<char>>& grid) {
        mp = grid;
        m = grid.size();
        n = grid[0].size();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (mp[i][j] == '1') {
                    res++;
                    dfs(i, j);
                }
            }
        }
        return res;
    }
    void dfs(int i, int j) {
        mp[i][j] = '0';
        for (const auto [dr, dc] : d) {
            int row = dr + i;
            int col = dc + j;
            if (row >= 0 && row < m && col >= 0 && col < n && mp[row][col] == '1') {
                dfs(row, col);
            }
        }
    }
};

File: Data Structures & Algorithms/swim-in-rising-water/submission-2.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        n = len(grid)
        d = [(0,1),(1,0),(-1,0),(0,-1)]
        heap = []
        heapq.heappush(heap, (grid[0][0], 0, 0))
        visited = set()
        res = 0
        while heap:
            t, r, c = heapq.heappop(heap)
            res = max(t, res)
            if r == n - 1 and c == n - 1:
                return res
            visited.add((r, c))

            for dr, dc in d:
                row = dr + r
                col = dc + c
                if row in range(n) and col in range(n):
                    if (row, col) in visited:
                        continue
                    heapq.heappush(heap, (grid[row][col], row, col))

        